src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.SkeletonReferenceCache.cs

```diff
@@ -248,15 +252,20 @@ private sealed class SkeletonReferenceSet
             private readonly DeferredDocumentationProvider _documentationProvider;
 
             /// <summary>
-            /// Use WeakReference so we don't keep MetadataReference's alive if they are not being consumed. 
-            /// Note: if the weak-reference is actually <see langword="null"/> (not that it points to null),
-            /// that means we know we were unable to generate a reference for those properties, and future
-            /// calls can early exit.
+            /// The actual assembly metadata produced from the data pointed to in <see cref="_storage"/>.
+            /// </summary>
+            private readonly AsyncLazy<AssemblyMetadata?> _metadata;
+
+            /// <summary>
+            /// Mapping from different values of <see cref="MetadataReferenceProperties"/> to the actual <see
+            /// cref="MetadataReference"/> produced for it.  Used so that if the same compilation is referenced from
+            /// different projects with/without the same properties that the right references are handed out (and shared
+            /// if possible).
             /// </summary>
             /// <remarks>
             /// This instance should be locked when being read/written.
             /// </remarks>
-            private readonly Dictionary<MetadataReferenceProperties, WeakReference<MetadataReference>?> _metadataReferences = new();
+            private readonly Dictionary<MetadataReferenceProperties, AsyncLazy<MetadataReference?>> _metadataReferences = new();
```

Comment:
>_metadataReferences [](http://example.com/codeflow?start=100&length=19)

Is there value in caching the MetadataReference objects? Those should be cheap to create.

Reply:
If you think it's cheap, i'm def willing to potentially remove this.  Slightly worried about the implication though... e.g. does anything in the compiler depend on reference-identity of these things?  Does the assembly manager care?  Etc.?

Reply:
The compiler shouldn't care. It cares about identity of AssemblyMetadata objects, but not references IIRC.

Reply:
ok.  will remove.

Reply:
removed now.

Reply:
That sure makes things a lot simpler -- thanks @tmat!